File: app/src/main/java/io/github/gusandrianos/foxforreddit/viewmodels/PostViewModelFactory.java
package io.github.gusandrianos.foxforreddit.viewmodels;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import io.github.gusandrianos.foxforreddit.data.repositories.PostRepository;

public class PostViewModelFactory extends ViewModelProvider.NewInstanceFactory {

    private PostRepository mPostRepository;

    public PostViewModelFactory(PostRepository postRepository) {
        mPostRepository = postRepository;
    }

    @SuppressWarnings("unchecked") //suppress return
    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        return ((T) new PostViewModel(mPostRepository));
    }
}

File: app/src/main/java/io/github/gusandrianos/foxforreddit/viewmodels/PostViewModelFactory.kt
package io.github.gusandrianos.foxforreddit.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider.NewInstanceFactory
import io.github.gusandrianos.foxforreddit.data.repositories.PostRepositoryKotlin

@Suppress("UNCHECKED_CAST")
class PostViewModelFactory(private val mPostRepository: PostRepositoryKotlin) : NewInstanceFactory() {
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        return PostViewModel(mPostRepository) as T
    }
}